File: api/misc/src/main/java/com/puresoltechnologies/commons/misc/ConfigurationParameter.java
package com.puresoltechnologies.commons.misc;

import java.util.Properties;

import com.puresoltechnologies.commons.math.LevelOfMeasurement;
import com.puresoltechnologies.commons.math.ParameterWithArbitraryUnit;

/**
 * This class specifies a single configuration parameter.
 * 
 * @author Rick-Rainer Ludwig
 * 
 * @param <T>
 *            is the type of the parameter value.
 */
public class ConfigurationParameter<T> extends ParameterWithArbitraryUnit<T> {

	private static final long serialVersionUID = 969219744450680004L;

	private final String propertyKey;
	private final T defaultValue;

	public ConfigurationParameter(String name, String unit,
			LevelOfMeasurement levelOfMeasurement, String description,
			Class<T> type, String propertyKey, T defaultValue) {
		super(name, unit, levelOfMeasurement, description, type);
		this.propertyKey = propertyKey;
		this.defaultValue = defaultValue;
	}

	/**
	 * This method returns a key name which can be used in {@link Properties} to
	 * store and retrieve the parameter value.
	 * 
	 * @return A {@link String} with the key name is returned.
	 */
	public final String getPropertyKey() {
		return propertyKey;
	}

	/**
	 * This method returns a default value of the configuration parameter. This
	 * value is to be used if no setting was made.
	 * 
	 * @return A value of T is returned containing the default value.
	 */
	public final T getDefaultValue() {
		return defaultValue;
	}

}

File: webui/webui/src/main/typescript/ui/configuration/ConfigurationParameter.ts
/**
 * This class represents a single configuration parameter. The whole definition is defined here.
 */
class ConfigurationParameter {

    /**
     * This method is used to convert a single configuration parameter into a 
     * ConfigurationParameter object.
     */
    public static fromJSON(
        preferencesGroup: PreferencesGroup,
        groupIdentifier: string,
        configurationParameter: any) {
        return new ConfigurationParameter(preferencesGroup,
            groupIdentifier,
            configurationParameter.name,
            configurationParameter.unit,
            configurationParameter.levelOfMeasurement,
            configurationParameter.description,
            configurationParameter.type,
            configurationParameter.propertyKey,
            configurationParameter.path,
            configurationParameter.defaultValue,
            String(configurationParameter.valueRepresentation));
    }

    constructor(
        public preferencesGroup: PreferencesGroup,
        public groupIdentifier: string,
        public name: string,
        public unit: string,
        public levelOfMeasurement: LevelOfMeasurement,
        public description: string,
        public type: string,
        public propertyKey: string,
        public path: string,
        public defaultValue: any,
        public valueRepresentation: string) {
    }

}

